google/cloud/internal/url_encode.cc

```diff
@@ -24,35 +25,68 @@ namespace internal {
 
 namespace {
 
-using UrlMappings =
-    std::array<std::pair<absl::string_view, absl::string_view>, 25>;
-
-constexpr auto kEncodeMappings = UrlMappings{{
-    {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-    {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-    {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-    {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-    {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
-}};
-
-// Until we require C++17, it isn't worth trying to eliminate the duplication
-// using a constexpr function to produce kDecodeMappings from kEncodeMappings.
-constexpr auto kDecodeMappings = UrlMappings{{
-    {"%20", " "}, {"%22", "\""}, {"%23", "#"},  {"%24", "$"}, {"%25", "%"},
-    {"%26", "&"}, {"%2B", "+"},  {"%2C", ","},  {"%2F", "/"}, {"%3A", ":"},
-    {"%3B", ";"}, {"%3C", "<"},  {"%3D", "="},  {"%3E", ">"}, {"%3F", "?"},
-    {"%40", "@"}, {"%5B", "["},  {"%5C", "\\"}, {"%5D", "]"}, {"%5E", "^"},
-    {"%60", "`"}, {"%7B", "{"},  {"%7C", "|"},  {"%7D", "}"}, {"%7F", "\177"},
-}};
+std::string ToHex(int i) {
+  std::stringstream stream;
+  stream << "%" << std::setfill('0') << std::setw(2) << std::uppercase
+         << std::hex << i;
+  return stream.str();
+}
+
+std::string Escape(char c) {
+  static auto const* const kDict = []() {
+    static std::unordered_map<char, std::string> m;
+    for (auto i = 0; i < 32; ++i) {
+      m[static_cast<char>(i)] = ToHex(i);
+    }
+    for (auto c : {
+             ' ', '\"', '#', '$', '%', '&',  '+', ',', '/', ':', ';', '<',
+             '=', '>',  '?', '@', '[', '\\', ']', '^', '`', '{', '|', '}',
+         }) {
+      m[c] = ToHex(static_cast<int>(c));
+    }
+    for (auto i = 127; i < 256; ++i) {
+      m[static_cast<char>(i)] = ToHex(i);
+    }
+    return &m;
+  }();
+  auto it = kDict->find(c);
+  if (it != kDict->end()) return it->second;
+  return std::string{c};
+}
+
+// Returns 0-15 if c is in [0-9A-Fa-f], and -1 otherwise.
+int ParseDigitHex(char c) {
+  if ('0' <= c && c <= '9') return c - '0';
+  if ('A' <= c && c <= 'F') return c - 'A' + 0xA;
+  if ('a' <= c && c <= 'f') return c - 'a' + 0xa;
+  return -1;
+}
 
 }  // namespace
 
 std::string UrlEncode(absl::string_view value) {
-  return absl::StrReplaceAll(value, kEncodeMappings);
+  std::string s;
+  for (auto c : value) {
+    absl::StrAppend(&s, Escape(c));
+  }
```

Comment:
But instead, consider getting rid of the map and doing every char on the fly, as well as doing conversions in place.  For example, ...
```
std::string UrlEncode(absl::string_view value) {
  std::string s;
  const char kDigits[] = "0123456789ABCDEF";
  for (unsigned char c : value) {
    if (WeNeedToEscape(c)) {
      s.push_back('%');
      s.push_back(kDigits[(c >> 4) & 0xf]);
      s.push_back(kDigits[c & 0xf]);
    } else {
      s.push_back(c);
    }
  }
  return s;
}
```

Reply:
Accepted, thanks.

Reply:
Apparently I used `char` and `unsigned char` interchangeably. Does it matter?